### ISSUE TITLE
Should we be allowing to rebindAll to pick up inherited properties?

### ISSUE THREAD

**Issue by matthewdunsdon** [d3fc/d3fc-rebind#12]
I have been exploring how d3-dispatch v1.0.0 works with d3fc-rebind and would like to be able to rebind the 'on' property in a component.

Since ``rebindAll`` uses ``Object.keys``, which does not include properties in the prototype chain, and ``myDispatchInstance.on`` is an inherited property it looks like I can not rebind the ``on`` without having to do some workaround in the consuming code.

Do we think we should be allowing to rebind and rebindAll to pick up inherited properties?

**Comment by matthewdunsdon** [comment]
I am not sure this is the best example, but here is a gist showing this issue:

https://gist.github.com/matthewdunsdon/1379c0c32ea0e1123ec69b41a174b0cc

**Comment by ColinEberhardt** [comment]
I agree, this is unexpected behaviour. Within d3fc we don't use prototype inheritance, however rebind should work regardless of the internal implementation of a component.

Do you fancy creating a PR?

**Comment by chrisprice** [comment]
Good spot. I'm not sure there's a good technique to allow this though. I
think we'll end up dragging in methods from higher up the prototype chain.
Can anyone propose an algorithm?

Two alternatives jump to mind -

The transforms are pluggable so just supply a custom one which rebinds
direct prototype methods.

Just use include("on") in this case (or the shorthand alias).

What do you think?

**Comment by matthewdunsdon** [comment]
:bulb: One idea I had was that we allow items on the prototype chain to be included by default.  With this is place, we could either:

1. Add a exclude prototype chain properties transform — _fc.**excludeAllInherited**()_
1. Add specific transforms for include to only use own properties — _fc.**includeOwn**(...names)_ 
1. Update existing include transform (so that only allows own properties) and add specific transforms to use prototype chain — _fc.**includeInherited**(...names)_

I may need to think about this more.

 :thought_balloon: _I am not sure how we would want exclude to play in the second or third case._

**Comment by chrisprice** [comment]
Sorry, I'd missed exactly what was causing the issue so my previous suggestions are invalid. 

Thinking it through, I have a few requirements for a solution -

1. As already noted, `rebind('x')` should work for any property named `x` no matter where it is located. 
2.  I would like to avoid complicating `rebindAll` for users following the standard non-prototype pattern as I think this will be the most common use case. I.e. don't make them explicitly exclude prototypical properties (e.g. `rebindAll(includeOwn/excludeInherited())`). 
3. I don't want to lose the ability to specify bespoke transforms e.g. `rebindAll(x => x.toLowerCase())`. 


Whilst I think any of @matthewdunsdon's suggestions would fix this issue, I think they'd all fall foul of 1. As a slight twist I'd propose pulling the include/exclude inherited functionality up to the top-level -

* Add new method `rebindDeep` which has similar functionality to `rebindAll` but additionally includes properties from the prototype chain.
* Modify `rebind` to skip the iteration (i.e. delegation to `rebindAll`) and directly create the rebound methods (`src/createReboundMethod.js` is already a standalone, probably left over from a refactor).

Any comments?

**Comment by chrisprice** [comment]
Actually thinking about #13, maybe we should just convert the whole thing to a component e.g. -

````js
rebind().keys(['foo', 'bar']).prefix('x').inherited(true)(target, source)
```

And add a shorthand -

```js
rebind('foo', 'bar')(target, source)

**Comment by chrisprice** [comment]
And after some more thinking, I'm not sure how my suggestion would handle the equivalent of `includeMap`.

**Comment by chrisprice** [comment]
For now I've implemented the fix to `rebind` (b5c7013) which fixes the reported issue (rebinding `on` from `d3.dispatch`). I'm not going to implement `rebindDeep` or similar as I don't want to bloat the API with another method if there isn't an actual issue to be solved.
